planner/core/logical_plan_builder.go

```diff
@@ -2474,3 +2488,34 @@ func getInnerFromParentheses(expr ast.ExprNode) ast.ExprNode {
 	}
 	return expr
 }
+
+// replaceVirtualColumn tries to replace a virtual generated column in given expression with its definition expression
+// stored in 'colMap'
+func replaceVirtualColumn(colMap map[int64]expression.Expression, expr expression.Expression) (expression.Expression, error) {
+	switch x := expr.(type) {
+	case *expression.Constant:
+		return x, nil
+	case *expression.Column:
+		if e, ok := colMap[x.ID]; ok {
+			return e, nil
+		}
+		return x, nil
+	case *expression.ScalarFunction:
+		args := make([]expression.Expression, 0, len(x.GetArgs()))
+		for _, arg := range x.GetArgs() {
+			r, err := replaceVirtualColumn(colMap, arg)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			args = append(args, r)
+		}
+		sf, err := expression.NewFunction(x.GetCtx(), x.FuncName.L, x.RetType, args...)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		return sf, nil
+	default:
+		// there should be no CorrelatedColumn in a data source
+		return nil, errors.Errorf("Unexpected data source column: %s", expr.String())
```

Comment:
Here you can use `errors.AddStack`, then you don't need to call `errors.Trace` in other places.